src/sources/kubernetes_logs/pod_metadata_annotator.rs

```diff
@@ -153,112 +170,235 @@ fn annotate_from_file_info(
-    }
+    let legacy_key = fields_spec
+        .container_name
+        .path
+        .as_ref()
+        .map(|k| &k.path)
+        .map(LegacyKey::Overwrite);
+
+    log_namespace.insert_source_metadata(
+        Config::NAME,
+        log,
+        legacy_key,
+        path!("container_name"),
+        file_info.container_name.to_owned(),
+    );
 }
 
-fn annotate_from_metadata(log: &mut LogEvent, fields_spec: &FieldsSpec, metadata: &ObjectMeta) {
-    for (key, val) in [
-        (&fields_spec.pod_name, &metadata.name),
-        (&fields_spec.pod_namespace, &metadata.namespace),
-        (&fields_spec.pod_uid, &metadata.uid),
+fn annotate_from_metadata(
+    log: &mut LogEvent,
+    fields_spec: &FieldsSpec,
+    metadata: &ObjectMeta,
+    log_namespace: LogNamespace,
+) {
+    for (legacy_key, metadata_key, value) in [
+        (&fields_spec.pod_name, path!("pod_name"), &metadata.name),
+        (
+            &fields_spec.pod_namespace,
+            path!("pod_namespace"),
+            &metadata.namespace,
+        ),
+        (&fields_spec.pod_uid, path!("pod_uid"), &metadata.uid),
     ]
     .iter()
     {
-        if let (Some(key), Some(val)) = (&key.path, val) {
-            log.insert(key, val.to_owned());
+        if let Some(value) = value {
+            let legacy_key = legacy_key
+                .path
+                .as_ref()
+                .map(|k| &k.path)
+                .map(LegacyKey::Overwrite);
+
+            log_namespace.insert_source_metadata(
+                Config::NAME,
+                log,
+                legacy_key,
+                *metadata_key,
+                value.to_owned(),
+            );
         }
     }
 
-    if let (Some(key), Some(owner_references)) =
-        (&fields_spec.pod_owner.path, &metadata.owner_references)
-    {
-        log.insert(
-            key,
+    if let Some(owner_references) = &metadata.owner_references {
+        let legacy_key = fields_spec
+            .pod_owner
+            .path
+            .as_ref()
+            .map(|k| &k.path)
+            .map(LegacyKey::Overwrite);
+
+        log_namespace.insert_source_metadata(
+            Config::NAME,
+            log,
+            legacy_key,
+            path!("pod_owner"),
             format!("{}/{}", owner_references[0].kind, owner_references[0].name),
-        );
+        )
     }
 
     if let Some(labels) = &metadata.labels {
-        if let Some(pod_label_prefix) = &fields_spec.pod_labels.path {
-            for (key, val) in labels.iter() {
-                let key_path = path!(key);
-                log.insert(
-                    (PathPrefix::Event, (&pod_label_prefix.path).concat(key_path)),
-                    val.to_owned(),
-                );
-            }
+        for (key, value) in labels.iter() {
+            let key_path = path!(key);
+            let legacy_key = fields_spec
+                .pod_labels
+                .path
+                .as_ref()
+                .map(|k| &k.path)
```

Comment:
pretty nit but it looks like all the way up to the `concat()` could be pulled out of the for loop

Reply:
I _think_ there was a compiler complaint here, but I'll double check and share the error if there was one.

Reply:
Resolved in 3161817cea1c1d85fd5d1050a877208c53db6bb2